### PR TITLE
Tickets/dm 30268

### PR DESCRIPTION
**Pull request by wvreeven**

### CODE REVIEW

**Review by r-owen — APPROVED** [review]
This looks nice.

I have a few suggestions I hope you will consider. The big one (which could be a later ticket) is to make an abstract base class for devices.

File: python/lsst/ts/envsensors/command_handler.py [hunk 1]
```diff
@@ -103,7 +103,7 @@ async def handle_command(self, command: str, **kwargs: typing.Any) -> None:
         self.log.info(f"Handling command {command} with kwargs {kwargs}")
         func = self.dispatch_dict[command]
         try:
-            await func(**kwargs)
+            await func(**kwargs)  # type: ignore
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/envsensors/command_handler.py:121]: Thanks for the suggestion. Done.

File: python/lsst/ts/envsensors/command_handler.py [hunk 2]
```diff
@@ -235,9 +235,9 @@ def _get_device(self, configured_device: dict) -> typing.Optional[typing.Any]:
         Raises
         ------
         RuntimeError
-            In case an incorrect _configuration has been loaded.
+            In case an incorrect configuration has been loaded.
         """
-        device = None
+        device: Any = None
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/envsensors/command_handler.py:328]: Sorry that I didn't remind you of this. Ticket DM-30072 will handle that.

File: python/lsst/ts/envsensors/rpi_serial_hat.py [hunk 3]
```diff
@@ -26,10 +26,11 @@
 
 from typing import Any, Dict
 import logging
-import RPi.GPIO as gpio
-import serial
+import RPi.GPIO as gpio  # type: ignore
+import serial  # type: ignore
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/envsensors/rpi_serial_hat.py:30]: That indeed is the reason.

File: python/lsst/ts/envsensors/rpi_serial_hat.py [hunk 4]
```diff
@@ -26,10 +26,11 @@
...
 import time
 from threading import RLock
+from typing import Tuple
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/envsensors/rpi_serial_hat.py:33]: I moved this import to line 27. The rest I will do in ticket DM-30072.

File: python/lsst/ts/envsensors/rpi_serial_hat.py [hunk 5]
```diff
@@ -209,7 +210,7 @@ def read_timeout(self, timeout: float) -> None:
     def terminator(self) -> str:
         """Serial data line terminator string ('str')."""
         self._message(
-            "Serial data line terminator string read: {}.".format(self._read_timeout)
+            "Serial data line terminator string read: {}.".format(self.read_timeout)
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/envsensors/rpi_serial_hat.py:213]: Sorry that I didn't remind you of this. Ticket DM-30072 will handle that.

File: python/lsst/ts/envsensors/rpi_serial_hat.py [hunk 6]
```diff
@@ -300,7 +301,7 @@ async def close(self) -> None:
                 else:
                     self._message("Serial port already closed.")
 
-    def readline(self) -> str:
+    def readline(self) -> Tuple[str, str, str]:  # type: ignore
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/envsensors/rpi_serial_hat.py:304]: Without it `mypy` complains about a missing return statement. Ticket DM-30072 will handle that.

File: python/lsst/ts/envsensors/command_handler.py [hunk 7]
```diff
@@ -104,12 +119,81 @@ async def handle_command(self, command: str, **kwargs: Any) -> None:
...
+
+        """
+        # KEY_DEVICES is mandatory.
+        if KEY_DEVICES not in configuration:
+            return False
+        # Only one key allowed.
+        if len(configuration.keys()) != 1:
+            return False
+        # Validate the device configurations.
+        device_configurations = configuration[KEY_DEVICES]  # type: ignore
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/envsensors/command_handler.py:149]: Ticket DM-30072 will handle that.

File: python/lsst/ts/envsensors/command_handler.py [hunk 8]
```diff
@@ -104,12 +119,81 @@ async def handle_command(self, command: str, **kwargs: Any) -> None:
...
+            return False
+        # Make sure that KEY_FTDI_ID is present for VAL_FTDI devices.
+        if device_configuration[KEY_TYPE] == VAL_FTDI:
+            if KEY_FTDI_ID not in device_configuration:
+                return False
+        # Make sure that KEY_SERIAL_PORT is present for VAL_SERIAL devices.
+        if device_configuration[KEY_TYPE] == VAL_SERIAL:
+            if KEY_SERIAL_PORT not in device_configuration:
+                return False
+        return True
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/envsensors/command_handler.py:194]: That is a great suggestion. No longer a boolean is returned but a CommandError is raised now.

File: python/lsst/ts/envsensors/constants.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+KEY_CHANNELS = "channels"
+KEY_DEVICES = "devices"
+KEY_FTDI_ID = "ftdi_id"
+KEY_NAME = "name"
+KEY_RESPONSE = "response"
+KEY_SERIAL_PORT = "serial_port"
+KEY_TELEMETRY = "telemetry"
+KEY_TYPE = "type"
+VAL_FTDI = "FTDI"
+VAL_SERIAL = "Serial"
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/envsensors/constants.py:51]: I was considering doing that and then didn't. Not sure anymore why. Done.

File: tests/test_command_handler.py [hunk 10]
```diff
@@ -23,100 +23,273 @@
...
+                    KEY_TYPE: VAL_SERIAL,
+                    "port": "ABC",
+                },
+            ]
+        }
+        self.assertFalse(
+            self.command_handler._validate_configuration(configuration=configuration)
+        )
+
+    def validate_response(self, response_code):
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply tests/test_command_handler.py:158]: Done

File: tests/test_command_handler.py [hunk 11]
```diff
@@ -23,100 +23,273 @@
...
+                    KEY_NAME: "Test1",
+                    KEY_CHANNELS: 4,
+                    KEY_TYPE: VAL_SERIAL,
+                    "port": "ABC",
+                },
+            ]
+        }
+        self.assertFalse(
+            self.command_handler._validate_configuration(configuration=configuration)
+        )
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply tests/test_command_handler.py:126]: Very good points, though the code presently is testing `configure` with one invalid configuration. I have merged the two tests and it is looping over several invalid configurations now.

**Pull request merged**